### ISSUE TITLE
AccessDenied: Access Denied when trying to .put or .get on S3

### ISSUE THREAD

**Issue by danmight** [aws-amplify/amplify-js#5058]
**Describe the bug**
When I try to get or put a file on S3 using Storage.put or Storage.get I get an access denied error.

Amplify CLI - 4.13.4
aws-amplify - 2.2.6

**To Reproduce**

I'm using React.

` const awsKey = await Storage.put(fileName, pdfBlob, {level: 'public', type: 'application/json'})`

I have a valid file name and blob. The interesting thing is this code worked before. I deleted my entire amplify instance and rebuilt it and that's when it stopped working. I was also getting a Client Not Authorized Error when I tried to use the IAM authRole  for authentication on API Gateway to call a lambda function. I got around that by making a cognito authorizer and using that as the auth method.

I'm wondering if I'm missing something in the rebuilding that's causing the IAM roles to not work or be invoked.. 

I checked that the IAM roles and the policies that Amplify attaches to the cognito authRole and they are there.

Under "Last Activity" on the authRole in the IAM console it shows no prior activity. It seems like these roles are not being used to authenticate my user to put or get. 

I noticed the same thing when I was struggling with the API Gateway IAM auth issue. I deleted the whole amplify project because I could only get the API Gateway to call a lambda function on my master branch and master environment. All other branch/environment combinations got not authorized error. When I rebuilt it, it stopped working on the master branch as well for the API Gateway, which leads me to believe there's some setting not right but I've checked all the obvious places for permissions and they are all the usual ones auto-generated by amplify that have worked before. 

Debugging Window:
<img width="1103" alt="Screen Shot 2020-03-06 at 4 03 07 PM" src="https://user-images.githubusercontent.com/47647812/76129611-fbbe4180-5fc4-11ea-988e-2e64cfbe79b0.png">

**Comment by danmight** [comment]
A follow up note to this, is that if I run the Policy Simulator it will return "allowed" when run on the buckets I'm trying to access, so the policy appears to be okay as well. I'm not sure why the calls are being denied when called through my app.

**Comment by danmight** [comment]
I was able to solve this. When ampify generated the federated identity pool it set the "Get Role" to "From Token". I'm not sure if that is correct or not, but if I switch it to "Default Role" and make sure the authRole is set correctly for authenticated users, it works. However, this still doesn't fix the API gateway auth issue from my other issue above. I still have to use custom cognito authorizers for that. For whatever reason IAM roles don't work for authentication there.

<img width="1103" alt="Screen Shot 2020-03-06 at 4 03 07 PM" src="https://user-images.githubusercontent.com/47647812/76167998-a30bb780-6130-11ea-82a0-3284aea6b9ba.png">

**Comment by ashika01** [comment]
@danmight could you provide some info on how to setup amplify project?

**Comment by danmight** [comment]
You can go ahead and close this. I don't have the console logs from when I set it up, but I don't think I chose anything crazy or did anything that didn't involved amplify <service> add and follow the questions it asks. I was able to at least figure out a solution to move forward, so that is good!

**Comment by ashika01** [comment]
@danmight Yeah, I have seen that certain times where something in the setup has been incorrect or something from cache messing around. As long as you got it going forward, its all good. If you have any other issues or need clarifications feel free to open up a new issue :)

**Comment by danielblignaut** [comment]
@danmight @ashika01  I've been battling this issue all day, think I figured out the gist of it. When IdentityProvider has choose from token and your user has custom group assigned to them, the role from their group overrides the default identityPool auth role... Therefore when performing operations to S3, the system is looking at your group role, which has no S3 policies attached.

**Comment by danielblignaut** [comment]
another hiccup when adding storage... if you want unauthorised users to have s3 access, "Enable access to unauthenticated identities" must be ticked in your IdentityPool which is not set to false in my instance (unsure if amplify auth cat lets you update this from the cli but it is listed as an option in the parameters.json file)

**Comment by danmight** [comment]
@danielblignaut Genius! Thanks for posting what your follow up. This just connected the dot on the last piece for me. I could not figure out why an early version I had worked fine. I introduced groups on sign up later, so my original users did not have any assigned and that explains why they were working. I hadn't actually written code to do anything with the groups yet, but they were getting assigned and probably getting assigned with no permissions. Well done.

**Comment by danielblignaut** [comment]
@danmight no problem... To fix the issue, I ended up having to do this nastiness inside the s3 cloudformation template... here's just a snippet from one of the policy's but you would need to update all of the "authenticated" policies (Would love to hear if this is advisable route or if there's a better potential solution as I'd have to remember to update this file every time I edit groups in my amplify setup):

```json
"S3AuthPublicPolicy": {
			"DependsOn": [
				"S3Bucket"
			],
			"Condition": "CreateAuthPublic",
			"Type": "AWS::IAM::Policy",
			"Properties": {
				"PolicyName": {
					"Ref": "s3PublicPolicy"
				},
				"Roles": {
					"Fn::Split": [
						",",
						{
							
							"Fn::Join": [
								",", [
									{
										"Ref": "authRoleName"
									},
									{
										"Fn::Sub": [
											"${deployName}-adminGroupRole,${deployName}-dietitianGroupRole,${deployName}-dietitianManagerGroupRole,${deployName}-userGroupRole",
											{
												"deployName": {
													"Ref": "authcognitoUsersUserPoolId"
												}
											}
										]
									}
								]
							]
						}
					]
				},
				"PolicyDocument": {
					"Version": "2012-10-17",
					"Statement": [
						{
							"Effect": "Allow",
							"Action": {
								"Fn::Split" : [ "," , {
									"Ref": "s3PermissionsAuthenticatedPublic"
								} ] 
							},
							"Resource": [
								{
									"Fn::Join": [
										"",
										[
											"arn:aws:s3:::",
											{
												"Ref": "S3Bucket"
											},
											"/public/*"
										]
									]
								}
							]
						}
					]
				}
			}
		}
```

You can see the horrible stuff happening in my roles array. The issue here is that you need to pass the role name (not ARN) into the roles array for assignment to your custom groups. However, you can't edit the outputs of the cognito user group's cloudformation as this file get's regenerated on every amplify push, overwriting your changes and the only outputs in that file are the role ARN not role name. Whilst you could hard code them into the s3 cloudformation, it would prevent your s3 storage working in the multiple environment setup in amplify, as each role name is prefixed with the environment's cognito user pool ID...

I'd suggest storage cat is reviewed for some of these mines and the role name is added to user group cloudformation template output or relook at that file being generated on every amplify push.

**Comment by ashika01** [comment]
@danmight @danielblignaut So the policy is not configured properly when you do configuration from `amplify add storage`?

**Comment by danmight** [comment]
@ashika01 It configures it correctly unless you have groups assigned incognito. For some reason, it is not assigning the authRole that cognito builds and configures correctly if you assign a user to a group in cognito.

**Comment by danielblignaut** [comment]
@ashika01 

**User role issue**

as @danmight mentioned... to reproduce, run the following:
1. add auth via amplify.
2. add a user group to cognito
3. add storage category
4. create a user and add to that group
5. login as user and attempt file upload

note that when logged in, the IAM role associated to the user is not {appname}-authRole, it is instead {cognito pool}-{user group name}GroupRole. This leads to an issue because the correct s3 permissions that are generated inside of the storage category's cloudformation template are associated only to {appname}-authRole or {appname}-unauthRole . I fixed this by using the snippet above however this could be refactored better. The reason why It's quite hacky is because the userPoolGroups cloudformation template can't be modified by the developer as it is generated on every amplify push and therefore changes are overwritten. My suggested change to this file was to add the role name's as outputs so that we can reference them directly to attach IAM policies to them. Unfortunately cloudformation only allows you to attach policy's by role name, not role ARN which is currently the only exported item.

**Guest S3 access issue**

if you want unauthorised users to have s3 access, "Enable access to unauthenticated identities" must be ticked in your IdentityPool which is not set to false in my instance (unsure if amplify auth cat lets you update this from the cli but it is listed as an option in the parameters.json file)...

see checkbox below.

<img width="1153" alt="Screenshot 2020-03-16 at 10 10 59" src="https://user-images.githubusercontent.com/1087177/76745875-7c700100-676e-11ea-9186-c31fb7e3b5a0.png">

as far as I can tell, the issue here is that without this box being checked, your guest user won't receive a jwt token for guest requests (amplify does this automatically when configuring cognito as far as i can tell)... therefore when attempting to file upload, you don't have an access token and your user will get an access denied issue. I believe that "Enable access to unauthenticated identities" is an auth category config as it's in the parameters.json file. Perhaps when adding storage, there needs to be a note to the user that this has to be activated and if the user prompts yes, this setting is automatically changed. 

**Amplify push - Existing schema attributes cannot be modified or deleted**

to reproduce,

1. add auth category to your project
2. add a custom schema attribute to your cognito cloudformation template
3. run amplify push
4. add storage category
5. run amplify push

in this instance, do a diff between the auth category cloudformation files from after step 2 v after step 4. you will notice that adding storage regenerates your auth cloudformation and has removed the custom schema attributes that you added.

**Comment by tammomueller** [comment]
@ashika01 Has there been any movement on this ticket. We are facing the same issues where we want to manage role-based access to resources in S3.

**Comment by ashika01** [comment]
@danielblignaut Looks like this one happen in the CLI Configuration phase. I will transfer it to the CLI team. They should be able to help you :)
